### ISSUE TITLE
Discriminative model training seems hardwired to Span Candidates

### ISSUE THREAD

**Issue by chaturv3di** [HazyResearch/snorkel#1062]
Currently, Snorkel only supports `span` objects as training candidates. In a previous issue [c.f. #1055], I figured out how to implement custom `CandidateSpace` classes to handle candidates that are entire documents or sentences. Having done that, I could proceed with writing LFs for these candidates and training marginals over label vectors.

Now I just ran into an issue while training discriminative models, which suggests that the assumption of `span` as candidate is hardwired into training of discriminative models. In particular, there's `_preprocess_data()` method in `snorkel/learning/pytorch/rnn/rnn_base.py` that in turn calls methods `get_word_start()` and `get_word_end()` that are only implemented in `TemporarySpan` class (and are not declared in `TemporaryContext`). 

Here's the offending code segment:
```python
for candidate in candidates:
    # Mark sentence
    args = [
        (candidate[0].get_word_start(), candidate[0].get_word_end(), 1),
        (candidate[1].get_word_start(), candidate[1].get_word_end(), 2)
    ]
```

**Moreover**, in the above code segment, it seems that a `candidate` is always a pair, which suggests that this code is hardwired to the `Spouse = candidate_subclass('Spouse', ['person1', 'person2'])` example. `candidate[0]` seems to be referring to `person1` and `candidate[1]` to `person2`. 

@ajratner, please correct me if I got this wrong. Otherwise, I'll now get down to implementing new classes that can handle entire documents as candidates.

**Comment by ajratner** [comment]
Hi @chaturv3di thanks for the great issue!

The model you're referring to---the PyTorch `rnn_base.py` implementation---is indeed set up with a `_preprocess_data()` method specifically for binary `Candidate` objects defined over spans.  However, you can modify this sub-function to extend the applicability of the model.  This is done, for example, in the TensorFlow implementation (the [TextRNN](https://github.com/HazyResearch/snorkel/blob/master/snorkel/learning/tensorflow/rnn/text_rnn.py) class, used in the [Crowdsourcing tutorial](https://github.com/HazyResearch/snorkel/blob/master/tutorials/crowdsourcing/Crowdsourced_Sentiment_Analysis.ipynb)), and should be fairly straightforward!

If you're interested in submitting a PR extending the PyTorch model similarly that would be great!  Thanks for raising either way!

In the meantime, closing the issue (and marking as Q&A for future reference), but happy to re-open
